File: homework2/steven_law/q1_SinglyLinkedList.cpp
// Time spent: 29 mins 35 seconds

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

// creates new Node with data val at front; returns head. O(1) time.
Node* insertAtFront(Node* head, int val) {

    Node* newNode = new Node();

    newNode -> data = val; 
    newNode -> next = head;
    head = newNode;

    return head;
}

// creates new Node with data val at end; returns head. O(n) time.
Node* insertAtBack(Node* head, int val) {

    Node* newNode = new Node();

    newNode -> data = val;

    if(head == nullptr){ // if head is null return newNode
        newNode -> next = nullptr;
        return newNode;
    }

    Node* temp = head;

    while(temp -> next != nullptr ){
        temp = temp->next;
    }

    temp->next = newNode;
    temp->next->next = nullptr;

    return head;
}

// creates new Node with data val after Node loc; returns head. O(1) time.
Node* insertAfter(Node* head, int val, Node* loc) {

    Node* newNode = new Node();                                                                                                                                                        
    newNode->data = val;

    newNode->next = loc->next;                                                                                                                                                         
    loc->next = newNode;        

    return head;
}

// creates new Node with data val before Node loc; returns head. O(n) time.
Node* insertBefore(Node* head, int val, Node* loc) {

    Node* newNode = new Node();
    newNode->data = val;

    Node* temp = head;

    if (head == loc) {                                                                                                                                                                 
      newNode->next = head;                                                                                                                                                          
      return newNode;                                                                                                                                                                
    } 

    while(temp -> next != loc ){
        temp = temp->next;
    }

    newNode->next = loc;
    temp->next = newNode;

    return head;
 
}

// removes first Node; returns head. O(1) time.
Node* deleteFront(Node* head) {
    if(head == nullptr){
        return head;
    }

    Node* temp = head;                                                                                                                                                                 
    head = head->next;

    delete temp;      //can't leak memory

    return head;
}

// removes last Node; returns head. O(n) time.
Node* deleteBack(Node* head) {
 
     if (head == nullptr) return nullptr; 
     
     if (head->next == nullptr) {                                                                                                                                                       
        delete head;                                                                                                                                                                   
        return nullptr;                                                                                                                                                                
    }                                                                                                                                                                                  
                                                                                                       
  Node* temp = head;        

  while (temp->next->next != nullptr) {                                                                
      temp = temp->next;                       
    }

  delete temp->next;
  temp->next = nullptr;

  return head;
}

// deletes Node loc; returns head. O(n) time.
Node* deleteNode(Node* head, Node* loc) {

  if (head == nullptr) return nullptr;

  if (head == loc) {                                                                                                                                                                 
      Node* newHead = head->next;                                                                      
      delete head;                                                                                                                                                                   
      return newHead;                                                                                                                                                                
  }  
    
    Node* temp = head;

    while(temp->next != loc){
        temp = temp->next;
    }

    Node* temp2 = temp->next->next;

    delete temp->next;
    temp->next = temp2;

    return head;
}

// returns length of the list. O(n) time.
int length(Node* head) {
    int count = 0;

    while(head != nullptr){
        head = head->next;
        count++;
    }

    return count;
}

// reverses the linked list iteratively. O(n) time.
Node* reverseIterative(Node* head) {
    if(head == nullptr || head->next == nullptr ) return head;
    

    Node* prev = nullptr;
    Node* curr = head;
    Node* temp = curr;

    while(curr != nullptr){
        temp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = temp;
    }

    return prev;
}

// reverses the linked list recursively (Hint: you will need a helper function.) O(n) time.
Node* reverseRecursive(Node* head) {
    return reverseHelper(head);
}

Node* reverseHelper(Node* curr) {  
    if (curr == nullptr || curr->next == nullptr) return curr;                                                                                                                     
                                                                                                                                                                                     
    Node* newHead = reverseHelper(curr->next);                                                       
    curr->next->next = curr;                                                                                                                                                       
    curr->next = nullptr;    
                                                                            
    return newHead;
}




int main() {

    return 0;
}

File: homework2/jeffery_owusu_agyemang/q1_SinglyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# O(1)
def insertAtFront(head, val):
    new_node = Node(val)
    new_node.next = head
    return new_node

# O(n)
def insertAtBack(head, val):
    new_node = Node(val)
    if not head:
        return new_node
    curr = head
    while curr.next:
        curr = curr.next
    curr.next = new_node
    return head

# O(1)
def insertAfter(head, val, loc):
    if not loc:
        return head
    new_node = Node(val)
    new_node.next = loc.next
    loc.next = new_node
    return head

# O(n)
def insertBefore(head, val, loc):
    if not head or not loc:
        return head
    if head == loc:
        return insertAtFront(head, val)
    prev = None
    curr = head
    while curr and curr != loc:
        prev = curr
        curr = curr.next
    if curr == loc:
        new_node = Node(val)
        prev.next = new_node
        new_node.next = curr
    return head

# O(1)
def deleteFront(head):
    if not head:
        return None
    return head.next

# O(n)
def deleteBack(head):
    if not head or not head.next:
        return None
    curr = head
    while curr.next.next:
        curr = curr.next
    curr.next = None
    return head

# O(n)
def deleteNode(head, loc):
    if not head or not loc:
        return head
    if head == loc:
        return head.next
    prev = None
    curr = head
    while curr and curr != loc:
        prev = curr
        curr = curr.next
    if curr == loc:
        prev.next = curr.next
    return head

# O(n)
def length(head):
    count = 0
    curr = head
    while curr:
        count += 1
        curr = curr.next
    return count

# O(n)
def reverseIterative(head):
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

# O(n)
def reverseRecursive(head):
    def helper(node, prev):
        if not node:
            return prev
        nxt = node.next
        node.next = prev
        return helper(nxt, node)
    return helper(head, None)